GraphicsContextGLANGLE::readPixelsImpl crashes on checkedProduct<size_t>

roberto-apple opened the pull request:
#### 3ee1b4aadb0f249d8b61d5cc3e02743eaadeda77
<pre>
GraphicsContextGLANGLE::readPixelsImpl crashes on checkedProduct&lt;size_t&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=289536">https://bugs.webkit.org/show_bug.cgi?id=289536</a>
<a href="https://rdar.apple.com/146461356">rdar://146461356</a>

Reviewed by Kimmo Kinnunen.

Align calls to GraphicsContextGLANGLE::readPixelsImpl().
Remove call to unsafeMakeSpan().

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::readPixelsImpl):

Canonical link: <a href="https://commits.webkit.org/292236@main">https://commits.webkit.org/292236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b82dbe3db31060d613f65afe9b83977429890291

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100370 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72699 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29964 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53028 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11083 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3784 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45165 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81251 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3877 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102407 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16323 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81715 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22621 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81089 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20294 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25666 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3080 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/15642 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22343 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27479 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22002 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25476 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

roberto-apple commented on LayoutTests/ipc/read-pixels-buffer-object-overflow.html:
Layout test is modification of crash.html provided in https://rdar.apple.com/146461356

kkinnunen-apple commented on LayoutTests/ipc/read-pixels-buffer-object-overflow.html:
```diff
@@ -0,0 +1,90 @@
+            }).finally(() => {
+                globalThis.testRunner?.notifyDone();
+            });
+        }, 100);
```
The test will crash future tests, since you don't cleanup the connections.
You can observe this by running the test with Debug build , run-webkit-tests --debug --repeat 1000.
It's better to not test this way.

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -570,7 +570,10 @@ void GraphicsContextGLANGLE::readPixelsBufferObject(IntRect rect, GCGLenum forma
 std::optional<IntSize> GraphicsContextGLANGLE::readPixelsImpl(IntRect rect, GCGLenum format, GCGLenum type, GCGLsizei bufSize, uint8_t* rawData, bool readingToPixelBufferObject)
 {
-    auto data = unsafeMakeSpan(rawData, checkedProduct<size_t>(rect.width(), rect.height(), 4));
+    auto product = checkedProduct<size_t>(rect.width(), rect.height(), 4);
```
The spans are already made by the callers, and this bug was added when the unsafeMakeSpan stuff was added.

Try to organise the code so that the signature becomes:
```
std::optional<IntSize> GraphicsContextGLANGLE::readPixelsImpl(IntRect rect, GCGLenum format, GCGLenum type, std::span<uint8_t> data)
```


Try if the PBO read path can construct zero-length std::span into the offset, and detect readiongToPixelBufferObject by that property (size() == 0).

kkinnunen-apple reviewed (commented):
For a test, add a test in 
TestGraphicsContextGLCocoa.mm

Add your test to read too large dimensional rect, and then add the expectation that it fails with gl error.

roberto-apple commented on Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
```diff
@@ -383,6 +383,120 @@ static bool hasMultipleGPUs()
     gl = nullptr;
 }
 
+TEST_F(GraphicsContextGLCocoaTest, readPixelsSuccess)
```
TODO: Move read pixels tests into test fixture to share common setup.

kkinnunen-apple replied:
Extremely nice logical thinking around the tests, e.g. to test a failure case you're first testing a positive case, ensuring that the tested function sometimes returns non-failing. Great!

roberto-apple commented on Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
```diff
@@ -383,6 +383,120 @@ static bool hasMultipleGPUs()
+
+    uint8_t gotValues[4] = { 0, 0, 0, 0 };
+    IntRect rect(1, 1, 1, 1);
+    context->readPixelsBufferObject(rect, GraphicsContextGL::RGBA, GraphicsContextGL::UNSIGNED_BYTE, (GCGLintptr)gotValues, 4, 0);
```
TODO: Test currently fails. Colors don't match and invalid operation error is returned.

kkinnunen-apple replied:
So the buffer object read is a different thing.
The caller creates a pixel buffer object and binds that. Read pixels is done with an offset. The offset is bytes into the pixel buffer object. And the read pixels result is written to the object. The client then maps the object and reads the results from that.
https://www.khronos.org/opengl/wiki/Pixel_Buffer_Object

kkinnunen-apple replied:
You can also punt testing PBO to land this change, if you want. The function needs a bit of refactoring, I can do also  that  add the test. The refactoring is related to moving code so that readPixels/readPixelsWithStatus would share code, but then readPixelsBufferObject would be its own hunk, so that we wouldn't need to construct an unsafe span.

roberto-apple replied:
Will remove pbo tests.

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -595,10 +594,8 @@ std::optional<IntSize> GraphicsContextGLANGLE::readPixelsImpl(IntRect rect, GCGL
 
 #if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
-    if (!readingToPixelBufferObject && !attrs.alpha && (format == GraphicsContextGL::RGBA || format == GraphicsContextGL::BGRA) && (type == GraphicsContextGL::UNSIGNED_BYTE) && (m_state.boundReadFBO == m_fbo || (attrs.antialias && m_state.boundReadFBO == m_multisampleFBO)))
+    if (data.size() > 0 && !attrs.alpha && (format == GraphicsContextGL::RGBA || format == GraphicsContextGL::BGRA) && (type == GraphicsContextGL::UNSIGNED_BYTE) && (m_state.boundReadFBO == m_fbo || (attrs.antialias && m_state.boundReadFBO == m_multisampleFBO)))
```
small nitpick, this could be `data.empty()`

roberto-apple committed:
GraphicsContextGLANGLE::readPixelsImpl crashes on checkedProduct<size_t>
https://bugs.webkit.org/show_bug.cgi?id=289536
rdar://146461356

Reviewed by Kimmo Kinnunen.

Align calls to GraphicsContextGLANGLE::readPixelsImpl().
Remove call to unsafeMakeSpan().

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::readPixelsImpl):

Canonical link: https://commits.webkit.org/292236@main
